bite-size-notes/week-4-d-Example of LQR-name.md

```diff
@@ -0,0 +1,350 @@
+Ptm(2,1) = PV(i,3);
+Ptm(2,2) = PV(i,4);
+K(i,:) = inv(R)*B'*Ptm;
+
+end
+
+
+[y2, dyx] = ode45(@(t,x) dyn(t,x,A,B),[0 (tf-ti)],x0,options);
+
+figure 
+plot(y2,dyx,'o','LineWidth',3)
+xlabel('time')
+ylabel('State')
+legend('uncontrol State x1' ,'uncontrol State x2')
+set(gca,'FontSize',20)
+
+% plot the control gain K
+nfig = nfig+1; figure(nfig)
+figure
+plot(tk,K,'r','LineWidth',3)
+xlabel('time')
+ylabel('Kalman gain K')
+set(gca,'FontSize',20)
+
+%return
+% simulate the sytem response using ode45
+options=odeset('RelTol',1e-10);
+[t,x]=ode45(@(t,x) xdiff(t,x,flag,A,B,tk,K),[0 (tf-ti)],x0,options);
+% plot the state x
+nfig = nfig+1; figure(nfig)
+figure
+plot(t,x,'b','LineWidth',3)
+xlabel('time')
+ylabel('State')
+legend('Controled State x1' ,'Controled State x2')
+set(gca,'FontSize',20)
+
+%return
+% find and plot the input u
+
+[m,n]=size(x);     % Length of time, x is m
+[mR,nR] = size(R); % number of inputs = mR
+Kt = interp1(tk,K,t); % interpolate the K matrix
+u = zeros(m,mR);    % initialize the input
+for jj=1:1:m
+    u(jj) = -Kt(jj,:)*(x(jj,:))';
+end
+
+figure
+plot(t,u,'r','LineWidth',3)
+xlabel('time')
+ylabel('Input u')
+set(gca,'FontSize',20)
+
+
+
+
+
+
+
+
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%%% Function to find the solution to Riccati Eq
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+function dpv = dpp(t,p,flag,A,B,Q,R,E,F)
+%
+%    Called by PSolve to compute the Riccati matrix P.
+%
+%    [t,GP] = ode45('gramdiff',[ti tf],zeros(NT,1),[],A,B,Q,R,E).
+%
+%    See also PSOLVE.
+%
+%
+% Finding the P matrix PM
+% 
+PM = zeros(2);
+PM(1,1) = p(1);
+PM(1,2) = p(2);
+PM(2,1) = p(3);
+PM(2,2) = p(4);
+%
+%
+% computing P matrix derivative 
+% Backward in time
+dp = A'*PM + PM*A -PM*E*PM +Q;
+%
+%
+% computing P vector derivative
+dpv = zeros(4,1);
+
+dpv(1,1) = dp(1,1) ;
+dpv(2,1) = dp(1,2) ;
+dpv(3,1) = dp(2,1) ;
+dpv(4,1) = dp(2,2) ;
+end
+
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%%% Function to find the solution to system with optimal control
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+function xdiff = xdiff(t,x,flag,A,B,tK,K)
+%
+% solving xdot = AX + Bu
+
+Kt = interp1(tK,K,t);
+ut = -Kt*x;
+xdiff = A*x +B*ut;
+end
+
+function dx = dyn(t,x,A,B)
+
+dx = A*x; 
+end
+```
+
+# Results: 
+First we can we plot the states without control, and find the system blows up, since the eigenvalue of the A matrix is on the RHP.
+![uncon](https://github.com/p8410077/AA548-spr2024/assets/11802603/b7168051-1b6c-4784-8c36-71f37404da7b)
+
+By inplementing LQR, the system converge to 0 in 1 second. 
+![image](https://github.com/p8410077/AA548-spr2024/assets/11802603/389fc3c2-0a2a-4066-815f-6815c59bd34f)
+
+# Conclusion: 
+1. We succesfully use ode45 to solve Ricatti equation and find the optimal control for continuous finite time problem.
+2. For other variation of the LQR need different technic to solve, for example: the infinit time problen is not a ODE, therefore can't use ode45.
```

Comment:
It may be helpful to write out the math for the infinite time horizon problem and show why it isn't an ODE, and maybe a line on what other approaches there can be approach such problems.

Reply:
Thanks, I specified that ode45 can't be use for discrete time equation